protocols/randhound/randhound.go

```diff
@@ -1,251 +1,1006 @@
+	rh.ServerIdxToGroupNum = make([]int, nodes)
+	rh.ServerIdxToGroupIdx = make([]int, nodes)
+
+	rh.i1s = make(map[int]*I1)
+	rh.i2s = make(map[int]*I2)
+	rh.r1s = make(map[int]*R1)
+	rh.r2s = make(map[int]*R2)
+	rh.polyCommit = make(map[int][]abstract.Point)
+	rh.secret = make(map[int][]int)
+	rh.chosenSecret = make(map[int][]int)
+
+	rh.Done = make(chan bool, 1)
+	rh.SecretReady = false
+
+	return nil
 }
 
-// Leader (=client) refers to the node which initiates the RandHound protocol,
-// moves it forward, and ultimately outputs the generated public randomness.
-type Leader struct {
-	rc      []byte                 // Leader's trustee-selection random value
-	rs      [][]byte               // Peers' trustee-selection random values
-	i1      *I1                    // I1 message sent to the peers
-	i2      *I2                    // I2 - " -
-	i3      *I3                    // I3 - " -
-	i4      *I4                    // I4 - " -
-	r1      map[uint32]*R1         // R1 messages received from the peers
-	r2      map[uint32]*R2         // R2 - " -
-	r3      map[uint32]*R3         // R3 - " -
-	r4      map[uint32]*R4         // R4 - " -
-	states  map[uint32]*poly.State // States for deals and responses from peers
-	invalid map[uint32]*[]uint32   // Map to mark invalid shares
-	Done    chan bool              // For signaling that a protocol run is finished
+// Start initiates the RandHound protocol run. The client pseudo-randomly
+// chooses the server grouping, forms an I1 message for each group, and sends
+// it to all servers of that group.
+func (rh *RandHound) Start() error {
+
+	var err error
+
+	// Set timestamp
+	rh.time = time.Now()
+
+	// Choose client randomness
+	hs := rh.Suite().Hash().Size()
+	rand := make([]byte, hs)
+	random.Stream.XORKeyStream(rand, rand)
+	rh.cliRand = rand
+
+	// Determine server grouping
+	rh.server, rh.key, err = rh.Shard(rand, rh.groups)
+	if err != nil {
+		return err
+	}
+
+	// Set some group parameters
+	for i, group := range rh.server {
+		rh.threshold[i] = 2 * len(group) / 3
+		rh.polyCommit[i] = make([]abstract.Point, len(group))
+		g := make([]int, len(group))
+		for j, server0 := range group {
+			s0 := server0.ServerIdentityIdx
+			rh.ServerIdxToGroupNum[s0] = i
+			rh.ServerIdxToGroupIdx[s0] = j
+			g[j] = s0
+		}
+		rh.group[i] = g
+	}
+
+	// Comptue session id
+	rh.sid, err = rh.sessionID(rh.nodes, rh.faulty, rh.purpose, rh.time, rh.cliRand, rh.threshold, rh.Public(), rh.key)
+	if err != nil {
+		return err
+	}
+
+	// Multicast first message to grouped servers
+	for i, group := range rh.server {
+
+		index := make([]uint32, len(group))
+		for j, server := range group {
+			index[j] = uint32(server.ServerIdentityIdx)
+		}
+
+		i1 := &I1{
+			SID:       rh.sid,
+			Threshold: rh.threshold[i],
+			Group:     index,
+			Key:       rh.key[i],
+		}
+
+		rh.mutex.Lock()
+
+		// Sign I1 and store signature in i1.Sig
+		if err := signSchnorr(rh.Suite(), rh.Private(), i1); err != nil {
+			return err
```

Comment:
I think @ineiti meant the mutex won't be released in the error case. Not sure if this is problematic but to be sure you could add a `Unlock` in this branch:
```
if err := signSchnorr(rh.Suite(), rh.Private(), i1); err != nil {
          rh.mutex.Unlock()
          return err
}
```